scipy/stats/_distn_infrastructure.py

```diff
@@ -1252,21 +1261,25 @@ def moment(self, n, *args, **kwds):
                 mdict = {'moments': {1: 'm', 2: 'v', 3: 'vs', 4: 'vk'}[n]}
             else:
                 mdict = {}
-            mu, mu2, g1, g2 = self._stats(*args, **mdict)
-        val = _moment_from_stats(n, mu, mu2, g1, g2, self._munp, args)
+            mu, mu2, g1, g2 = self._stats(*goodargs, **mdict)
+        val = _moment_from_stats(n, mu, mu2, g1, g2, self._munp, goodargs)
 
         # Convert to transformed  X = L + S*Y
         # E[X^n] = E[(L+S*Y)^n] = L^n sum(comb(n, k)*(S/L)^k E[Y^k], k=0...n)
-        if loc == 0:
-            return scale**n * val
-        else:
-            result = 0
-            fac = float(scale) / float(loc)
-            for k in range(n):
-                valk = _moment_from_stats(k, mu, mu2, g1, g2, self._munp, args)
-                result += comb(n, k, exact=True)*(fac**k) * valk
-            result += fac**n * val
-            return result * loc**n
+        fac = _lazywhere(cond1_, (goodloc, goodscale),
+                         lambda loc, scale: scale / loc, self.badvalue)
+        for k in range(n):
+            valk = _moment_from_stats(k, mu, mu2, g1, g2, self._munp,
+                                      goodargs)
+            result[cond0] += comb(n, k, exact=True)*fac**k * valk
+        result[cond0] += fac**n * val
+        result[cond0] *= goodloc**n
+        res = goodscale**n * val
```

Comment:
Apparently, this implementation fails when `(loc==0) & (scale <= 0)`. For instance, the following calls fail:

```python
from scipy.stats import norm

norm.moment(2, 0, -4)
norm.moment(2, [1, 0, 2], [3, -4, -5])
```

Will try to address this and add some tests in the next commit